htlcswitch/link.go

```diff
@@ -2509,18 +2513,30 @@ func (l *channelLink) CheckHtlcForward(payHash [32]byte,
 	// Next, using the amount of the incoming HTLC, we'll calculate the
 	// expected fee this incoming HTLC must carry in order to satisfy the
 	// constraints of the outgoing link.
-	expectedFee := ExpectedFee(policy, amtToForward)
+	expectedFee := int64(ExpectedFee(policy, amtToForward))
+
+	// Calculate inbound fee that we charge based on the incoming amount.
+	// Add this to the outbound fee separately. Do not attempt to first
+	// create a combined fee by adding up the inbound and outbound base and
+	// proportional fees. Rounding may cause the total fee to be higher and
+	// potentially fail forwards for senders that use a different
+	// calculation.
+	inFee := inboundFee.CalcFee(amtToForward)
+
+	expectedFee += inFee
 
 	// If the actual fee is less than our expected fee, then we'll reject
 	// this HTLC as it didn't provide a sufficient amount of fees, or the
 	// values have been tampered with, or the send used incorrect/dated
 	// information to construct the forwarding information for this hop. In
 	// any case, we'll cancel this HTLC.
-	actualFee := incomingHtlcAmt - amtToForward
-	if incomingHtlcAmt < amtToForward || actualFee < expectedFee {
+	actualFee := int64(incomingHtlcAmt - amtToForward)
```

Comment:
`incomingHtlcAmt` and `amtToForward` are both uint32. So if `amtToForward` is greater than `incomingHtlcAmt` the actualFee would overflow. Maybe cast both values to int64 and then substract.

Reply:
Yes, good point. Fixed.

Although I think the end result is the same. Even if you subtract two unsigned ints and only then cast, the result isn't any different. Again because of the encoding on a lower level.